csrc/abstract_tensor_schedule.cpp

```diff
@@ -0,0 +1,261 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+
+#include <abstract_tensor.h>
+#include <abstract_tensor_schedule.h>
+#include <ir/internal_base_nodes.h>
+#include <iter_visitor.h>
+#include <val_graph.h>
+
+namespace nvfuser {
+
+namespace {
+
+class AbstractTensorSchedule {
+ public:
+  static void apply(
+      const AbstractTensor& abstract,
+      TensorView* concrete,
+      ValGraph* graph) {
+    AbstractTensorSchedule ats(abstract, concrete, graph);
+    ats.run();
+  }
+
+ private:
+  AbstractTensorSchedule(
+      const AbstractTensor& abstract,
+      TensorView* concrete,
+      ValGraph* graph)
+      : abstract_(abstract), concrete_(concrete), graph_(graph) {}
+
+  void run() {
+    findNearestProducers();
+
+    // Now, for each ValGroup in abstract, find the closest producer ValGroups
+    // with entries in concrete_ids_ and replay the path from them. If none
+    // exists, then do not include this dimension in the output
+    std::vector<IterDomain*> loop_domain;
+    for (const AbstractId& abs_id : abstract_.domain) {
+      IterDomain* new_id = replayAbstractId(abs_id);
+      if (new_id == nullptr) {
+        continue;
+      }
+      loop_domain.push_back(new_id);
+    }
+    concrete_->setLoopDomain(loop_domain);
+  }
+
+  ValGroup abstractIdToValGroup(const AbstractId& abs_id) {
+    if (abs_id.is<IterDomain*>()) {
+      return graph_->toGroup(abs_id.as<IterDomain*>());
```

Comment:
I originally thought maybe we could make this work for `IterDomain` abstract IDs as well as `ValGroupAndItsGraph`s. However, it's confusing to use this way so I might assert here that `abs_id.is<ValGroupAndItsGraph>()` instead. To see why, consider how you might use this functionality:
```c++
  AbstractTensor abten({tv2->axis(0), tv1->axis(0), tv1->axis(1)});
  abten.merge(0);
  abten.merge(0);
  abten.split(0, 128);
  IdModel id_model(&fusion);
  ValGraph& graph = id_model.idGraph(IdMappingMode::PERMISSIVE);
  applyAbstractSchedule(abten, tv, &graph);
```
The reason we need to build the graph _after_ scheduling `abten` is that the split and merge ops on `AbstractTensor` introduce new `IterDomain`s and `Merge`/`Split` expressions. However, those are actually disconnected from the `Fusion` and they are ignored when building the `ValGraph`. By contrast, if we start with `ValGroup`s in `abten`, then the `AbstractTensor::merge` and `AbstractTensor::split` methods update the graph so that those disconnected IDs and exprs are added to the appropriate (pre-existing) `ValGroup`s. I don't see a good way around this, so for now it is probably best to only support the case where all the `AbstractId`s in `abten` are `ValGroupAndItsGraph`s.

Reply:
Another option would be to fix up `IdModel` to consider "unused" `IterDomain`s and their expressions when building graphs. Currently we use `fusion->exprs()` and look at the `IterDomain`s in their TVs only, but this would let us grab the "unused" `IterDomain` resulting from `AbstractTensor::merge`, for example, as a `ValGroup` when needed. cc @naoyam.

Reply:
If it's clearly defined which domains and exprs should be considered, then I suppose it would be relatively simple. We shouldn't include truly dead domains and exprs, though.

Reply:
I tried to make this work but given the complexity of using `IterDomain`s instead of `ValGroup`s for this purpose, I removed this branch so we now require using `ValGroup`s in the `AbstractTensor`. We could try and enable it in the future if needed.